dali/pipeline/operators/arithmetic/arithmetic.h

```diff
@@ -0,0 +1,268 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_PIPELINE_OPERATORS_ARITHMETIC_ARITHMETIC_H_
+#define DALI_PIPELINE_OPERATORS_ARITHMETIC_ARITHMETIC_H_
+
+#include <memory>
+#include <string>
+#include <tuple>
+#include <vector>
+
+#include "dali/core/static_switch.h"
+#include "dali/kernels/tensor_shape.h"
+#include "dali/kernels/tensor_shape_print.h"
+#include "dali/kernels/type_tag.h"
+#include "dali/pipeline/operators/arithmetic/arithmetic_meta.h"
+#include "dali/pipeline/operators/arithmetic/expression_impl_factory.h"
+#include "dali/pipeline/operators/operator.h"
+
+namespace dali {
+
+/**
+ * @brief Cover every sample fully.
+ */
+inline std::tuple<std::vector<TileDesc>, std::vector<TileRange>> GetCover(
+    const kernels::TensorListShape<> &shape, int num_tasks) {
+  std::vector<TileDesc> descs;
+  std::vector<TileRange> ranges(num_tasks, {0, 0});
+  int samples_per_task = shape.num_samples() / num_tasks;
+  if (!samples_per_task)
+    samples_per_task = 1;
+  int prev_task_idx = -1;
+  for (int sample_idx = 0; sample_idx < shape.num_samples(); sample_idx++) {
+    int task_idx = sample_idx / samples_per_task;
+    descs.push_back({sample_idx, 0, task_idx, shape[sample_idx].num_elements(),
+                     shape[sample_idx].num_elements()});
+    if (task_idx != prev_task_idx) {
+      ranges[task_idx].begin = descs.size() - 1;
+      ranges[task_idx].end = descs.size();
+    } else {
+      ranges[task_idx].end = descs.size();
+    }
+    prev_task_idx = task_idx;
+  }
+  return std::make_tuple(descs, ranges);
+}
+
+/**
+ * @brief Divide the shape into groups of linear extents, separated evenly between threads
+ */
+inline std::tuple<std::vector<TileDesc>, std::vector<TileRange>> GetTiledCover(
+    const kernels::TensorListShape<> &shape, int extent, int num_tasks) {
+  Index total_elements = shape.num_elements();
+  std::vector<TileDesc> descs;
+  std::vector<TileRange> ranges(num_tasks, {0, 0});
+  Index elements_per_thread = total_elements / num_tasks;
+  // One extent covers all
+  if (total_elements < extent || elements_per_thread == 0) {
+    return GetCover(shape, num_tasks);
+  }
+  int task_idx = 0;
+  int prev_task_idx = -1;
+  std::vector<Index> covered_by_thread(num_tasks, 0);
+  for (int sample_idx = 0; sample_idx < shape.num_samples(); sample_idx++) {
+    int extent_idx = 0;
+    for (Index covered = 0; covered < shape[sample_idx].num_elements(); covered += extent) {
```

Comment:
suggestion: Increment `extent_idx` here instead of in line 95
```suggestion
    for (Index covered = 0; covered < shape[sample_idx].num_elements(); covered += extent, extent_idx++) {
```

Reply:
The line gets too long, I don't want to have to break `for`.

Reply:
Then store `shape[sample_idx].num_elements()` in a variable?